homework/index.js

```diff
@@ -30,17 +15,134 @@
     return elem;
   }
 
-  function main(url) {
-    fetchJSON(url, (err, repositories) => {
-      const root = document.getElementById('root');
-      if (err) {
-        createAndAppend('div', root, { text: err.message, class: 'alert-error' });
-        return;
-      }
-      createAndAppend('pre', root, { text: JSON.stringify(repositories, null, 2) });
+  function fetchJSON(url) {
+    return new Promise((resolve, reject) => {
+      const xhr = new XMLHttpRequest();
+      xhr.open('GET', url);
+      xhr.responseType = 'json';
+      xhr.onload = () => {
+        if (xhr.status >= 200 && xhr.status <= 299) {
+          resolve(xhr.response);
+        } else {
+          reject(new Error(`Network error: ${xhr.status} - ${xhr.statusText}`));
+        }
+      };
+      xhr.onerror = () => reject(new Error('Network request failed'));
+      xhr.send();
+    });
+  }
+
+  function loader(parent) {
+    const loaderDiv = createAndAppend('div', parent, { class: 'loaderDiv' });
+    createAndAppend('div', loaderDiv, { class: 'loader' });
+  }
+
+  function renderError(error) {
+    const root = document.getElementById('root');
+    createAndAppend('h1', root, { text: error.message });
+  }
+
+  function addRow(tableBody, description, className, elementContent) {
+    const tableRow = createAndAppend('tr', tableBody, { class: 'flex-div' });
+    createAndAppend('td', tableRow, {
+      text: description,
+      class: className,
+    });
+    createAndAppend('td', tableRow, {
+      text: elementContent,
     });
+    return tableRow;
   }
 
-  const HYF_REPOS_URL = 'https://api.github.com/orgs/HackYourFuture/repos?per_page=100';
-  window.onload = () => main(HYF_REPOS_URL);
+  function renderRepository(tableDiv, repository) {
+    tableDiv.innerHTML = '';
+
+    const table = createAndAppend('table', tableDiv);
+    const tbody = createAndAppend('tbody', table);
+    const firstRow = addRow(tbody, 'Repository:', 'td-header', ' ');
+    createAndAppend('a', firstRow.lastChild, {
+      text: repository.name,
+      href: repository.html_url,
+      target: '_blank',
+    });
+    if (repository.description) {
+      addRow(tbody, 'Description:', 'description-td td-header', repository.description);
+    }
+    addRow(tbody, 'Forks:', 'td-header', repository.forks_count);
+    addRow(tbody, 'Updated:', 'td-header', new Date(repository.updated_at).toLocaleString());
+  }
+
+  function renderContributions(url, contributesDiv) {
+    fetchJSON(url)
+      .then(contributors => {
+        contributesDiv.innerHTML = '';
+        createAndAppend('h1', contributesDiv, {
+          class: 'contributes-header',
+          text: 'Contributions',
+        });
+        const ul = createAndAppend('ul', contributesDiv);
+
+        Object.keys(contributors).forEach(contributor => {
```

Comment:
This is something I overlooked in the week 1 homework: the `contributors` obtained from `fetchJSON` is an array. While an array is also an object, its keys are simply the index values from zero up to `length-1`. So `Object.keys()` is not really useful here. Better use `.forEach()`:

```js
contributors.forEach(contributor => {
```

`contributor` is no longer a key or an index, but an element of the `contributors` array, which is a contributor object.